include/zephyr/drivers/stepper.h

```diff
@@ -0,0 +1,434 @@
+ *
+ * @see stepper_enable_constant_velocity_mode() for details.
+ */
+typedef int (*stepper_enable_constant_velocity_mode_t)(const struct device *dev,
+						       const enum stepper_direction direction,
+						       const uint32_t value);
+
+/**
+ * @brief Stepper Motor Controller API
+ */
+__subsystem struct stepper_api {
+	stepper_enable_t enable;
+	stepper_move_t move;
+	stepper_set_max_velocity_t set_max_velocity;
+	stepper_set_micro_step_res_t set_micro_step_res;
+	stepper_get_micro_step_res_t get_micro_step_res;
+	stepper_set_actual_position_t set_actual_position;
+	stepper_get_actual_position_t get_actual_position;
+	stepper_set_target_position_t set_target_position;
+	stepper_is_moving_t is_moving;
+	stepper_enable_constant_velocity_mode_t enable_constant_velocity_mode;
+};
+
+/**
+ * @endcond
+ */
+
+/**
+ * @brief Enable or Disable Motor Controller
+ *
+ * @param dev pointer to the stepper motor controller instance
+ * @param enable Input enable or disable motor controller
+ *
+ * @retval <0 Error during Enabling
+ * @retval  0 Success
+ */
+__syscall int stepper_enable(const struct device *dev, const bool enable);
+
+static inline int z_impl_stepper_enable(const struct device *dev, const bool enable)
+{
+	const struct stepper_api *api = (const struct stepper_api *)dev->api;
+
+	return api->enable(dev, enable);
+}
+
+/**
+ * @brief Set the micro_steps to be moved from the current position i.e. relative movement
+ *
+ * @param dev pointer to the stepper motor controller instance
+ * @param micro_steps target micro_steps to be moved from the current position
+ * @param async Pointer to a valid and ready to be signaled struct
+ *              k_poll_signal. (Note: if NULL this function will not notify
+ *              the end of the transaction, and whether it went successfully
+ *              or not).
+ *
+ * @retval <0 Error during Enabling
+ * @retval  0 Success
+ */
+__syscall int stepper_move(const struct device *dev, int32_t micro_steps,
+			   struct k_poll_signal *async);
+
+static inline int z_impl_stepper_move(const struct device *dev, const int32_t micro_steps,
+				      struct k_poll_signal *async)
+{
+	const struct stepper_api *api = (const struct stepper_api *)dev->api;
+
+	return api->move(dev, micro_steps, async);
+}
+
+/**
+ * @brief Set the target velocity to be reached by the motor
+ *
+ * @details For controllers such as DRV8825 where you
+ * toggle the STEP Pin, the pulse_length would have to be calculated based on this parameter in the
+ * driver. For controllers where velocity can be set, this parameter corresponds to max_velocity
+ * @note Setting max velocity does not set the motor into motion, a combination of set_max_velocity
+ * and move is required to set the motor into motion.
+ *
+ * @param dev pointer to the stepper motor controller instance
+ * @param micro_steps_per_second speed in micro_steps per second
+ *
+ * @retval <0 Error during Enabling
```

Comment:
In general, I'd expect a more explicit behavior at an interface level, that is, document what impl should return for what, eg, -EINVAL or -ENOTSUP if given velocity is not supported. This, in theory, allows writing more generic layers on top, as they can handle errors consistently.